posts/2022-01-03-kryo-pitfalls.adoc

```diff
@@ -0,0 +1,289 @@
+project are high speed, low size, and an easy to use API. The project is useful any time objects need
+to be persisted, whether to a file, database, or over the network.
+
+Kryo can also perform automatic deep and shallow copying/cloning. This is direct copying from object to
+object, not object to bytes to object.
+____
+
+This means that Kryo's use case is, just like other serialization libraries, sharing data between parties
+(i.e. between a writer and a reader). An important difference with other libraries is that we don't have
+to define a schema before we can start sending data (e.g. with Avro, we use
+https://avro.apache.org/docs/current/idl.html[Avro IDL]), and this can be a source of problems if it's not
+handled carefully.
+
+=== Registration
+
+While working with Kryo, we can _optionally_ register the classes that we want to serialize:
+
+[source,scala]
+----
+case class Person(id: Long, name: String, age: Int)
+val alice = Person(1L, "Alice", 42)
+
+val kryo = new Kryo()
+kryo.register(classOf[Person])
+
+// Write
+val output = new Output(new FileOutputStream("file.bin"))
+kryo.writeObject(output, alice)
+output.close()
+
+// Read
+val input = new Input(new FileInputStream("file.bin"))
+val aliceFromFile = kryo.readObject(input, classOf[Person])
+input.close()
+----
+
+*Note*: to make the previous snippet work, you have to add a _no-args_ constructor to `Person`. I decided
+to omit it for brevity, but it's required for every class that you serialize with Kryo.
+
+As mentioned, class registration is optional, but required by default, which can be disabled with
+`kryo.setRegistrationRequired(false)`.
+
+If we analyze the output from both procedures we find them identical:
+
+[source,shell]
+----
+❯ xxd object-with-registration.bin
+00000000: 0154 0201 416c 6963 e5                   .T..Alic.
+
+❯ xxd object-without-registration.bin
+00000000: 0154 0201 416c 6963 e5                   .T..Alic.
+----
+
+== Pitfall #1: Compatibility
+
+Initially one could be led to believe that there is no difference between registering and not registering
+a class, but let's take a look at the next snippet:
+
+[source,scala]
+----
+case class Message(wholePayload: String, size: Int, ts: Long)  // <- New Message class
+case class Person(id: Long, name: String, age: Int)
+val alice = Person(1L, "Alice", 42)
+
+val kryo = new Kryo()
+kryo.register(classOf[Person])
+kryo.register(classOf[Message])
+
+val output = new Output(new FileOutputStream("file.bin"))
+kryo.writeObject(output, alice)                                // <--- Write a Person
+output.close()
+
+val input = new Input(new FileInputStream("file.bin"))
+val messageFromFile = kryo.readObject(input, classOf[Message]) // <--- Read a Message
+input.close()
+----
+
+This is perfectly valid, even if `Message` and `Person` are completely different, they have different field
+names. The only similarity is that both classes have three fields with the same types, just with a different
```

Comment:
The first sentence seems to put the reader on the wrong foot: yes, the case classes are completely different as their field names are different.

```suggestion
This is perfectly valid, even if `Message` and `Person` are completely different as they have different field
names. The only similarity is that both classes have three fields with the same types, just with a different
```